sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,133 @@
+    this.intervalInMs = Number(options.intervalInMs);
+    this.client = options.client;
+    this.initialResponse = options.initialResponse;
+    this.previousResponse = options.previousResponse;
+    this.requestOptions = options.requestOptions;
+    this.stopped = Boolean(options.stopped);
+
+    if (options.state) this._state = options.state;
+    if (this.manual) return;
+    this.poll();
+  }
+
+  protected abstract getStateFromResponse(
+    response: HttpOperationResponse
+  ): LongRunningOperationStates;
+
+  protected abstract async sendRequest(options?: RequestOptionsBase): Promise<HttpOperationResponse>;
+
+  protected async processResponse(response: HttpOperationResponse): Promise<void | Result> {
+    if (!this.initialResponse) this.initialResponse = response;
+    this.previousResponse = response;
+    this.emit("newResponse", response, this);
+    this.state = this.getStateFromResponse(response);
+    if (this.isDone()) {
+      const result = await this.finalRequest();
+      this.emit("done", result);
+      return result;
+    }
+  }
+
+  protected async poll(): Promise<void> {
+    if (this.manual) return;
+    try {
+      if (!this.initialResponse) {
+        await this.processResponse(await this.initialRequest(this.requestOptions));
+      }
+      while (!this.isDone()) {
+        const interval: number = this.intervalInMs || this.getInterval();
+        await delay(interval);
+        await this.processResponse(await this.sendRequest());
+      }
+    } catch(e) {
+      this.emit("pollError", e, this);
+      throw e;
+    }
+  }
+
+  public abstract async initialRequest(options?: RequestOptionsBase): Promise<HttpOperationResponse>;
+
+  public abstract async finalRequest(options?: RequestOptionsBase): Promise<Result>;
+
+  public abstract async cancelRequest(options?: RequestOptionsBase): Promise<void>;
+
+  public abstract isDone(): boolean;
+
+  public set state(newState: LongRunningOperationStates) {
+    const newState = this.state !== newState;
+    this._state = newState;
+    if (newState) {
+      this.emit("newState", newState, this);
+    }
+  }
+
+  public get state(): LongRunningOperationStates {
+    return this._state;
+  }
+
+  public getInterval(): number {
+    return this.intervalInMs;
+  }
+ 
+  public async pollOnce(options?: RequestOptionsBase): Promise<void | Result> {
+    if (!this.manual) throw new Error("Manual retries are disabled on this poller");
+    try {
+      if (!this.initialResponse) {
+        await this.processResponse(await this.initialRequest(this.requestOptions));
+      }
+      return await this.processResponse(await this.sendRequest());
+    } catch (e) {
+      this.emit("pollError", e, this);
+      throw e;
+    }
+  }
+
+  public done(): Promise<void | Result | Error> {
+    return new Promise((resolve, reject) => {
+      this.on("done", resolve);
+      this.on("pollError", reject);
+    });
+  }
+
+  public stop(): void {
+    this.stopped = true;
```

Comment:
`this.stopped` never gets read and this call has no side effects.

Reply:
Added this.stopped as a condition to the while loop inside of "poll", as follows: `while (!this.stopped || !this.isDone()) {`

Reply:
I think we should also consider if we want to abort the last request or just ignore the response. 

I'm leaning towards the latter. If so, we'd have to check if stopped was called before updating state or notifying of done/error.

This is where having a state machine would be handy, since we could say 'stopped' is a final state and completes the `done()` promise.

Reply:
I'm not sure if I understood this. Stop only stops the while loop, like "forgetting" to loop. Just for the record, the main difference with "cancel" is that "cancel is an actual operation, while "stop" doesn't update the state at all. Another way to see this is that "stop" makes the poller "manual". In fact, we could merge both properties. Having said that, what do you think? What am I missing? 🌞

Reply:
This is a good discussion point. I understand that `cancel` stops the operation on the server, whereas `stop` is about not polling. If we wanted to make this more explicit I think we could rename cancel to `cancelOperation` and `stop` to `stopPolling`.

That aside, the real question I have is what does `stop` really imply? I think there are many possibilities:

1. Don't schedule another poll operation
2. Don't emit any more events
3. Complete all pending promises
4. Ignore in-flight requests that return

I think we are all in agreement about the first point, but the rest are open to debate.

Naively as a user, I expect `stop()` to indicate that I no longer wish to receive any updates about this long running operation. It would feel weird for me to call `stop()` and then sometime asynchronously later have whatever code I connected to the poller get called.

I think it's a good question about if a stopped poller can still be manually updated or not. Do we see this as a common use case? When would a developer initially want an automatic poller but then decide to switch to manual polling before the poller has completed?

If this is a pattern we want to support/encourage, maybe we should consider renaming `stop` to `pause` and having a `start` open to resume automatic polling. Alternatively, we could rename `stop` to `makeManual` if the switch was not reversible.